### PR TITLE
add GPTSAN model

### PR DESCRIPTION
**Pull request by tanreinama**
# Model description

GPTSAN is a Japanese language model using Switch Transformer. It has the same structure as the model introduced as Prefix LM in the T5 paper, and works with both Test Generation and Masked Language Model.

To add this model to the transformer, I did the following:
Porting GPTSAN to PyTorch. Model conversion. Creating model cards in HuggingFace Hub. Porting generation code.
The model card has already been uploaded. (https://huggingface.co/Tanrei/GPTSAN-japanese/)

Tokenizer uses GPT-NeoX-Japanese, and only new vocabulary files are uploaded to the model card. Minor differences are absorbed within the generation algorithm in the model's source code.

GPTSAN repository is:
https://github.com/tanreinama/GPTSAN

Discussion of HuggingFace integration is:
https://github.com/tanreinama/GPTSAN/issues/2

Thanks to: @ArthurZucker and @younesbelkada

### CODE REVIEW

**Review by ArthurZucker — COMMENTED** [review]
Hey! Thanks a lot for working on this! It's very cool to have the first Switch Transformer from the community! 
My main comment is that we should make the model compatible with our `generate()` function. Some specific arguments (maybe hybrid inputs etc) can be controled via the `generation_config` that was recently introduced!

File: src/transformers/models/gptsan_japanese/modeling_gptsan_japanese.py [hunk 1]
```diff
@@ -0,0 +1,1169 @@
...
+
+        """
+        hidden_states = self.wi(hidden_states)
+        hidden_states = self.act(hidden_states)
+        hidden_states = self.dropout(hidden_states)
+        hidden_states = self.wo(hidden_states)
+        return hidden_states
+
+
+class GPTSANJapaneseTop1Router(nn.Module):
```
[the remark this answers is not in the document]
Comment (tanreinama) [reply src/transformers/models/gptsan_japanese/modeling_gptsan_japanese.py:117]: It's a little different because of ```if self.training and self.jitter_noise > 0:``` line, so commit test didn't pass with ```copied from statement```.
However, since we can set jitter_noise to 0, so I respect the original and restore it, adding ```copied from```.

File: src/transformers/models/gptsan_japanese/modeling_gptsan_japanese.py [hunk 2]
```diff
@@ -0,0 +1,1169 @@
...
+class GPTSANJapaneseModel(GPTSANJapanesePreTrainedModel):
+    _keys_to_ignore_on_load_missing = [
+        r"wte.weight",
+    ]
+
+    def __init__(self, config: GPTSANJapaneseConfig):
+        super().__init__(config)
+        self.wte = nn.Embedding(config.vocab_size, config.d_model)
+        self.wpe = nn.Embedding(config.num_contexts, config.d_model)
+        self.wob = nn.Parameter(torch.zeros([config.vocab_size]))
```
[the remark this answers is not in the document]
Comment (tanreinama) [reply src/transformers/models/gptsan_japanese/modeling_gptsan_japanese.py:971]: Maybe this is carried over from GPT2
https://github.com/huggingface/transformers/blob/main/src/transformers/models/gpt2/modeling_gpt2.py

**Comment by tanreinama** [comment]
generate() now works with greedy_gen_mode, but I want contrastive_search to be the default. Is there any reference code somewhere for that?

**Comment by younesbelkada** [comment]
Yes, contrastive search should be supported in `transformers` but I think you need to tweak the caching mechanism. Maybe @gante can help here as I am not really sure 🙏

**Comment by tanreinama** [comment]
ok. I was confused about contrastive_search. This will work almost fine.
```
model.config.use_cache=True
model.config.do_sample=True
c = model.generate(x_tok, logits_processor=LogitsProcessorList([TopKLogitsWarper(120)]))
```

I would like to override _get_logits_processor and add TopKLogitsWarper to the default logs_processor.
```
logits_processor = super()._get_logits_processor(...)
if generation_config.top_k is not None:
    logits_processor.append(TopKLogitsWarper(generation_config.top_k))
return logits_processor
```

There was also a misunderstanding about the caching mechanism. I thought that cache saves everything up to the last time, and that SequenceLength is 1 every time forward is called, but it seems that's not the case. I can make it compatible.

**Comment by gante** [comment]
@tanreinama @younesbelkada contrastive search _should_ work out of the box if the model uses the usual caching mechanism. Prefix LM models are not the case, sadly (it's probably the same issue as GIT, which is also a prefix LM model) 😅 

I'd suggest to skip contrastive search for now, and to fix it in a subsequent PR (skip = skip tests and override `contrastive_search` such that an informative exception is thrown). I should be able to give better advice after I see what's happening with GIT :)

**Comment by tanreinama** [comment]
@ArthurZucker @younesbelkada
I have committed some updates in response to your comments.

In unit tests, the `wav2vec2_with_lm` module is causing errors. Is this due to conflicts? I didn't touch the this package...

**Comment by ArthurZucker** [comment]
The failing test is not related to you! If you pull from main it might get resolved! However you have `FAILED tests/models/gptsan_japanese/test_modeling_gptsan_japanese.py::GPTSANJapaneseForConditionalGenerationTest::test_logits - Failed: Timeout >120.0s` which means either the test should be marked as `#slow` or there is an issue int his test 👍🏻

**Comment by tanreinama** [comment]
I sync and pull from main so it closed automatically. I'll create new PR from merged code. thx,
